### ISSUE TITLE
This is an abomination

### ISSUE THREAD

**Issue by ashton314** [thaliaarchi/bfcoq#1]
Thalia… why would you do this?

**Comment by thaliaarchi** [comment]
For fun :)

Well I started with formalizing Whitespace as wscoq. I got the idea just before the final presentations in CS 401R, so I decided to present as well, even though I didn't need to as the TA. It works pretty well for expressing the semantics of the language, but the unstructured control flow makes the execution relation more clumsy and optimizations more difficult to perform without a control-flow graph and SSA form. I may revisit it later, once I've implemented the Relooper algorithm from Emscripten for structuring control flow for another holiday project to prep as the TA of CS 601R.

I decided to switch to Brainfuck, when someone else's presentation was to formalize Brainfuck and it seemed much easier to perform optimizations, since loops are structured. The optimizations are modest so far: only combining adjacent sequences of `<` and `>`, while preserving tape underflow, and combining adjacent sequences of `+` and `-`, but I've proven it sound, except for some details with normalization that I'm still working out. I have to redo a bunch of my lemmas, though, because yesterday I switched from the built-in byte type with 256 variants to a modular-arithmetic-based byte type from CompCert. I did that, so it'll hopefully be easier to prove properties, for example, that loops that add an odd integer to the cell will always reach zero and can be reduced to `Const 0`.
